### DISCUSSION TITLE
Smooth camera + MonoGame.Extended (tiled): issues with gaps between tiles at specific camera positions

### DISCUSSION THREAD

**Discussion by skeets23** [MonoGame/MonoGame#8587] [Graphics]
When having a "smooth" camera that doesn't lock on to the character and instead smoothly follows the character around, at specific camera positions, gaps form between all the tiles rendered by the tiled extension in MonoGame.Extended. (either in the X axis or the Y axis)

I've seen a lot of discussions on how to fix this, but they are usually quite dated and don't really seem conclusive.

In 2024, is there a good way to fix this?

I've tried:

- Rounding the camera to the nearest pixel (and the pixel size is determined by 1/zoom). This kind of works, but makes the camera have a subtle jitter?
- Setting SamplerState.PointClamp (this seems to make no difference, I'm assuming this was the default value)

**Comment by KelsamGames** [comment]
Let your camera have a floating point value, but round it when rendering (do not store the rounded value to the variable). That way it can mathematically move at sub-pixel increments, but will always draw at exact pixels.

**Comment by skeets23** [comment]
Thanks, but I double checked, and it turns out I was already doing that.

Hmmm, after further inspection, it's not the *camera* that is jittering, but the location of the *player* within the camera.

I've cleaned up my code a bit so I can paste it here:

```
 float pixelSize = 1 / _zoom;

 _camera.LookAt(player.position);

 if (cameraPreviousPosition == new Vector2())
 {
     cameraPreviousPosition = _camera.Position;
 }

 _camera.Position = Vector2.Lerp(cameraPreviousPosition, _camera.Position, deltaTime * 5);

 cameraPreviousPosition = _camera.Position;

 float newCameraX = (float)Math.Round(_camera.Position.X / pixelSize) * pixelSize;
 float newCameraY = (float)Math.Round(_camera.Position.Y / pixelSize) * pixelSize;

 _camera.Position = new Vector2(newCameraX, newCameraY);
```

I think the problem lies in the way the camera speed changes depending on how far away the camera is from the player. Which, at first makes me think the problem might be unavoidable... *However*, when I comment out the last line above, the jittering goes away 100%, but that brings back the gap between tiles. So the rounding is related somehow.

Ideally, I'd like the player to appear like they are "in place" when moving at a constant speed for more than a couple seconds, and that's mostly working, but the player "jitters" in place. So to be correct, the camera itself isn't actually jittering, but the position of the player within the camera is.

In summary, removing either of these 2 lines fixes the jittering (but the compromise either way isn't worth it):

`_camera.Position = Vector2.Lerp(cameraPreviousPosition, _camera.Position, deltaTime * 5);`
(but removing this line removes the "smooth camera" affect and just glues the camera to the player)

` _camera.Position = new Vector2(newCameraX, newCameraY);`
(but this re-introduces the gap between tiles problem)

Is there any other way to eliminate this "jittering"?

**Reply by KelsamGames** [reply]
Do not set `_camera.Position` with a rounded value, because on the next frame it'll modify the now-rounded value instead of the more accurate floating-point value that you want. Instead, just pass `Vector2.Round(_camera.Position)` in whatever rendering code where you are using that value to draw (i.e. use rounded value to draw, but keep the floating-point accuracy when mathing the value, never actually save/set/store the rounded value into a long-living variable).

**Reply by skeets23** [reply]
While the rounded value is technically being saved to a long-living variable, the rounded value is not used to calculate the next camera position. You'll see at the top of my code block:

>  _camera.LookAt(player.position);

So, the camera position is overwritten immediately, the rounded value is discarded at the start of the update loop. Only the `cameraPreviousPosition` value is used (which is the camera position before the rounding occurs).

This way, I don't need to apply rounding every single time I reference the camera coordinates to draw something. Unless I'm missing something, I don't think this would be a problem... (but I'm still pretty new to MonoGame)

**Reply by KelsamGames** [reply]
`LookAt` changes the viewing angle, not the position. You're setting the position via `_camera.Position = Vector2.Lerp(cameraPreviousPosition, _camera.Position, deltaTime * 5);` which is re-using its `.Position` value.

Disclaimer: as far as I am aware. I do not use Extended/Tiled so I am basing that on my experience with 3D matrix camera references. Either way, if not this exact value/scenario, something similar to this is almost certainly your underlying issue.

**Reply by skeets23** [reply]
Oh, I forgot to mention that I'm using a 2D `OrthographicCamera`, in which case, the `LookAt` function does in fact set the position. Here's the full `LookAt` function source code:

```
public override void LookAt(Vector2 position)
{
    Position = position - new Vector2((float)_viewportAdapter.VirtualWidth / 2f, (float)_viewportAdapter.VirtualHeight / 2f);
}
```

Just to be sure, I tried adding this to the top of the update function: `_camera.Position = new Vector2(0, 0);`, and the end result was the same 👍

It seems to be the variable camera speed (the Lerp function) in combination with the rounding that is causing the player position on the screen to jitter slightly despite moving at a constant speed.

**Reply by KelsamGames** [reply]
Hard to say much more without diving into all your actual code. I'd recommend watching vars from frame to frame to see which one(s) are the problem here. Perhaps you can use the lerp and just pass a rounded player position to the LookAt function or some such. Sorry I can't really be more help on this one, I don't use 2D ortho cameras.

**Reply by skeets23** [reply]
After a lot of thought, I think what's happening is this:

- Using `Lerp` means that the further away the player is from the center of the camera, the faster the camera moves.
- When the camera speed is faster, it "catches up" with the player, but that in turn makes the speed of the camera slower in the next frame (because there is not as much distance from the player), creating the jittery effect.

In the end, I think this would be a very difficult thing to get "just right", but I may have come across something that is "good enough"...

I added a "max distance" from the player, so when the camera gets too far away from the player, instead of just using `Lerp` to bring it closer, it keeps the player exactly `x` pixels away from the center (it will allow the player to get closer to the center, but not further). So, when the player is moving very quickly, the jitteriness disappears entirely.

So there is still some cases of jitteriness when moving at slower (or even normal) speeds, but the faster speeds were much more noticeable, so the jitter may be subtle enough at this point to be acceptable (and it only happens at some very exact speeds)

Anyways, thanks so much for taking the time to help me think through this. If the above gives you any brilliant ideas on how to fix the subtle jitter that still persists, I'd love to hear it, but otherwise, I think I'm ready to call this good enough. 😀

**Reply by skeets23** [reply]
@KelsamGames - Just thought you might want to know I actually solved this finally. I have the final solution below. Thanks again for the help.

**Comment by skeets23** [comment]
I definitively fixed this today. It probably took me way more time than it should have, but I'll share my solution in case it helps.

- Move the camera by sub-pixel increments, whatever you need to do for your smooth camera
- Check where the player will be rendered on the screen based on this sub-pixel camera position (remember, the rendering engine will automatically round the position on screen as it cannot render it at sub-pixel positions)
- Round the camera to avoid floating point issues with tile sets
- Check where the player will be render after the rounding (again, remember it will always be rendered at whole number coordinates even if it's technically at a sub-pixel location)
- If there's a difference, move the camera by the difference (it should be a whole number), so that the player will be rendered exactly where it would have been with the smooth camera before it was rounded
- Hooray! Smooth camera with whole pixel coordinates complete, no jittering or tile set clipping / artifacts at all 😀
